### PR TITLE
feat: uds config validation

### LINKED ISSUE #552: Config schema validation
### Overview
Currently there is no error output for unexpected config fields. It would be nice if there were a validation step that warned the user if a config was botched (yaml spacing, unexpected keys, schema misalignment etc).

For example, if a user forgets to add a `variables` key:
```
core:
  KEYCLOAK_CONFIG_IMAGE: nooooooope
```

Obviously no values would be passed to the core package.. but currently the deploy appears to continue to execute "successfully"

### PR DESCRIPTION
**Pull request by decleaver**
## Description

Adds validation for unmarshalling uds-config files

## Related Issue

Fixes #552 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-cli/blob/main/CONTRIBUTING.md) followed

### CODE REVIEW

File: src/test/e2e/commands_test.go [hunk 1]
```diff
@@ -112,6 +112,12 @@ func deploy(t *testing.T, tarballPath string) (stdout string, stderr string) {
 	return stdout, stderr
 }
 
+func deployWithError(t *testing.T, tarballPath string) (stdout string, stderr string) {
```
Comment (UncleGedd) [inline src/test/e2e/commands_test.go:115]:
```suggestion
func deployWithError(_ *testing.T, tarballPath string) (stdout string, stderr string) {
```

**Comment by UncleGedd** [comment]
Just a couple of merge conflicts to resolve and then I think this one is g2g

File: src/test/bundles/07-helm-overrides/uds-config-invalid.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+options:
+  log_level: debug
```
Comment (UncleGedd) [inline src/test/bundles/07-helm-overrides/uds-config-invalid.yaml:2]: We probably can't do this for `variables`, but is it possible to have it error out if `log_level` is misspelled?

Comment (UncleGedd) [reply src/test/bundles/07-helm-overrides/uds-config-invalid.yaml:2]: Noting that it does error out if `options` is misspelled which is 🤌

Comment (UncleGedd) [reply src/test/bundles/07-helm-overrides/uds-config-invalid.yaml:2]: I think that list is all the options, it'd be better to have that in a Go type though. I think today all those options exist in https://github.com/defenseunicorns/uds-cli/blob/main/src/cmd/root.go

Comment (decleaver) [reply src/test/bundles/07-helm-overrides/uds-config-invalid.yaml]: Added an "enum" for valid options and refactored function for unit testing

File: src/cmd/uds.go [hunk 3]
```diff
@@ -287,3 +286,19 @@ func chooseBundle(args []string) string {
...
+func UnmarshalAndValidateConfig(configFile []byte, bundleCfg *types.BundleConfig) error {
+	// read relevant config into DeployOpts.Variables
+	// need to use goyaml because Viper doesn't preserve case: https://github.com/spf13/viper/issues/1014
+	err := goyaml.UnmarshalWithOptions(configFile, &bundleCfg.DeployOpts, goyaml.Strict())
+	if err != nil {
+		return err
+	}
+	// validate config options
+	for optionName := range bundleCfg.DeployOpts.Options {
+		if !types.IsValidConfigOption(optionName) {
```
Comment (UncleGedd) [inline src/cmd/uds.go:299]: I think this is only used in the `cmd` pkg, can we move it there and make it private?

File: src/cmd/uds.go [hunk 4]
```diff
@@ -193,9 +194,7 @@ func loadViperConfig() error {
 		return err
 	}
 
-	// read relevant config into DeployOpts.Variables
-	// need to use goyaml because Viper doesn't preserve case: https://github.com/spf13/viper/issues/1014
-	err = goyaml.UnmarshalWithOptions(configFile, &bundleCfg.DeployOpts, goyaml.Strict())
+	err = UnmarshalAndValidateConfig(configFile, &bundleCfg)
```
Comment (UncleGedd) [inline src/cmd/uds.go:197]: Can this function be private?

File: src/test/bundles/07-helm-overrides/uds-config-invalid.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+options:
+  log_levelx: debug
+
+# helm-overrides is an invalid top level key. Missing `variables` parent key.
```
Comment (UncleGedd) [inline src/test/bundles/07-helm-overrides/uds-config-invalid.yaml:4]:
```suggestion
```

**Review by UncleGedd — APPROVED** [review]

**Review by UncleGedd — APPROVED** [review]

**Pull request merged**